### PR TITLE
feat: Misconfigured apps

### PR DESCRIPTION
**Pull request by jachym-tousek-keboola**
Jira: https://keboola.atlassian.net/browse/PSGO-408

**Changes:**
- Changed behavior of misconfigured apps to error 403. We can customize the page later.
- Fixed code review comments from previous PRs.

-----------

### CODE REVIEW

File: internal/pkg/service/appproxy/http/router.go [hunk 1]
```diff
@@ -95,65 +96,87 @@ func (r *Router) CreateHandler() http.Handler {
 	return mux
 }
 
-func (r *Router) createDataAppHandler(app DataApp) (http.Handler, error) {
+func (r *Router) createForbiddenHandler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		w.WriteHeader(http.StatusForbidden)
+		fmt.Fprintln(w, "Application has misconfigured OAuth2 provider.")
+	})
+}
```
Comment (michaljurecko) [inline internal/pkg/service/appproxy/http/router.go:104]: `forbiddenHandler` -> `configErrorHandler`?  forbidden sounds to me like it's the user's fault

File: internal/pkg/service/appproxy/http/router_test.go [hunk 2]
```diff
@@ -63,6 +63,20 @@ func TestAppProxyRouter(t *testing.T) {
 				assert.Equal(t, `Application "unknown" not found.`, string(body))
 			},
 		},
+		{
+			name: "broken-app",
+			run: func(t *testing.T, client *http.Client, m []*mockoidc.MockOIDC, appServer *appServer) {
+				// Request to unknown app
```
Comment (michaljurecko) [inline internal/pkg/service/appproxy/http/router_test.go:69]: `// Request to unknown app` -> `// Request to broken app`.

**Review by michaljurecko — APPROVED** [review]
LGTM

File: internal/pkg/service/appproxy/http/router.go [hunk 3]
```diff
@@ -95,65 +96,87 @@ func (r *Router) CreateHandler() http.Handler {
...
+	return r.protectedAppHandler(ctx, app)
 }
 
-func (r *Router) publicAppHandler(app DataApp) (http.Handler, error) {
+func (r *Router) publicAppHandler(ctx context.Context, app DataApp) http.Handler {
 	target, err := url.Parse("http://" + app.UpstreamHost)
 	if err != nil {
-		return nil, errors.Errorf(`cannot parse upstream url "%s" for app %s: %w`, app.UpstreamHost, app.ID, err)
+		r.logger.Errorf(ctx, `cannot parse upstream url "%s" for app <app> %s: %w`, app.UpstreamHost, app.Name, err.Error())
+		return r.configErrorHandler
```
Comment (michaljurecko) [inline internal/pkg/service/appproxy/http/router.go:119]: In future, we will need probably to include error details in the error page (at least an `exceptionId`), so the handler won't be same.

**Review by michaljurecko — APPROVED** [review]
LGTM

File: internal/pkg/service/appproxy/http/router.go (hunk 3 above)
Comment (michaljurecko) [reply internal/pkg/service/appproxy/http/router.go:119]: See `httpserver/error.go` and `WithExceptionID` interfaces for example.

Error in the `go-client` implements `WithExceptionID` interface.

image

Comment (michaljurecko) [reply internal/pkg/service/appproxy/http/router.go:119]: So you should:
- Generate an exeception ID for the error.
- Log the error, with details.
- In the handler, always return the same `exceptionID`.
- So we (support) can find it in the logs, and get error details.
- Logged message and message for the user may differ.

**Review by hosekpeter — APPROVED** [review]

**Pull request merged**